tensorly/backend/core.py

```diff
@@ -0,0 +1,854 @@
+import inspect
+import importlib
+import os
+import sys
+import threading
+import types
+import warnings
+
+import numpy as np
+import scipy.linalg
+import scipy.sparse.linalg
+
+
+_STATE = threading.local()
+_DEFAULT_BACKEND = None  # changed in `_initialize_default_backend`
+_KNOWN_BACKENDS = ('numpy', 'mxnet', 'pytorch', 'tensorflow', 'cupy')
+_LOADED_BACKENDS = {}
+
+
+def _initialize_default_backend():
+    """Setup the initial Tensorly backend"""
+    default = 'numpy'
+    backend = os.environ.get('TENSORLY_BACKEND', default)
+
+    if backend not in _KNOWN_BACKENDS:
+        msg = ("TENSORLY_BACKEND should be one of {%s}, got %r. Defaulting to "
+               "'%r'") % (', '.join(map(repr, _KNOWN_BACKENDS)), backend, default)
+        warnings.warn(msg, UserWarning)
+        backend = default
+
+    set_backend(backend, make_default=True)
+
+
+class set_backend(object):
+    """Sets the backend for TensorLy.
+
+    Can be used to set the backend globally, or locally as a contextmanager.
+
+    Parameters
+    ----------
+    backend : {'numpy', 'mxnet', 'pytorch', 'tensorflow', 'cupy'}
+        The name of the backend to use. Default is 'numpy'.
+    make_default : bool, optional
+        If True, the backend will become the default backend for all threads.
+        Note that this only affects threads where the backend hasn't already
+        been explicitly set. If False (default) the backend is only set
+        contextually for the local thread.
+
+    Examples
+    --------
+    Set the backend to numpy globally for this thread:
+
+    >>> import tensorly as tl
+    >>> tl.set_backend('numpy')
+
+    Set the backend to numpy globally for all threads:
+
+    >>> tl.set_backend('numpy', make_default=True)
+
+    Set the backend to use pytorch inside a context:
+
+    >>> with tl.set_backend('pytorch'):  # doctest: +SKIP
+    ...     pass
+    """
+    def __init__(self, backend, make_default=False):
+        if not isinstance(backend, Backend):
+            if backend not in _LOADED_BACKENDS:
+                # load the backend
+                if backend in _KNOWN_BACKENDS:
+                    importlib.import_module('tensorly.backend.%s_backend' % backend)
+                else:
+                    msg = ("Unknown backend %r, known backends are {%s}"
+                           % (backend, ', '.join(map(repr, _KNOWN_BACKENDS))))
+                    raise ValueError(msg)
+            backend = _LOADED_BACKENDS[backend]
+
+        # Set the backend
+        self._old_backend = getattr(_STATE, 'backend', None)
+        self._new_backend = _STATE.backend = backend
```

Comment:
What about making these class attributes?
If we want separation between the context and the set_backend function (not sure we do), we could also have set_backend as a staticmethod..

Reply:
I see this being a class as an implementation detail, and would prefer to keep all attributes on this object private. Users should use `get_backend` to query the state, not look at the contextmanager directly.

Reply:
I am suggesting *class attributes*, which *can* be private.

We could have a BackendManager class which would store the state, backend, etc and implement staticmethods which could then be called by various functions.

e.g. 
```python
class BackendManager():
    _old_backend = ...
    _all_backend_info = ...

    def __init__(self, ...):
        pass
```